pkg/kubelet/cm/node_allocatable.go

```diff
@@ -0,0 +1,216 @@
+				glog.Error(message)
+				// TODO: Set Limits to be a few bytes greater than current memory usage to try to force kernel to reclaim memory.
+				// Retry after a minute.
+				time.Sleep(time.Minute)
+			}
+		}(cm.nodeInfo, cm.recorder)
+	}
+	// Now apply kube reserved and system reserved limits if required.
+	if nc.EnforceNodeAllocatable.Has(SystemReservedEnforcementKey) {
+		glog.V(2).Infof("Enforcing system reserved on cgroup %q with limits: %+v", nc.SystemReservedCgroupName, nc.SystemReserved)
+		if err := enforceExistingCgroup(cm.cgroupManager, nc.SystemReservedCgroupName, nc.SystemReserved); err != nil {
+			return fmt.Errorf("failed to enforce System Reserved Cgroup Limits: %v", err)
+		}
+	}
+	if nc.EnforceNodeAllocatable.Has(KubeReservedEnforcementKey) {
+		glog.V(2).Infof("Enforcing kube reserved on cgroup %q with limits: %+v", nc.KubeReservedCgroupName, nc.KubeReserved)
+		if err := enforceExistingCgroup(cm.cgroupManager, nc.KubeReservedCgroupName, nc.KubeReserved); err != nil {
+			return fmt.Errorf("failed to enforce Kube Reserved Cgroup Limits: %v", err)
+		}
+	}
+	return nil
+}
+
+// enforceExistingCgroup updates the limits `rl` on existing cgroup `cName` using `cgroupManager` interface.
+func enforceExistingCgroup(cgroupManager CgroupManager, cName string, rl v1.ResourceList) error {
+	cgroupConfig := &CgroupConfig{
+		Name:               CgroupName(cName),
+		ResourceParameters: getCgroupConfig(rl),
+	}
+	glog.V(4).Infof("Enforcing limits on cgroup %q with %d cpu shares and %d bytes of memory", cName, cgroupConfig.ResourceParameters.CpuShares, cgroupConfig.ResourceParameters.Memory)
+	if !cgroupManager.Exists(cgroupConfig.Name) {
+		return fmt.Errorf("%q cgroup does not exist", cgroupConfig.Name)
+	}
+	if err := cgroupManager.Update(cgroupConfig); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Returns a ResourceConfig object that can be used to create or update cgroups via CgroupManager interface.
+func getCgroupConfig(rl v1.ResourceList) *ResourceConfig {
+	// TODO(vishh): Set CPU Quota if necessary.
+	if rl == nil {
+		return nil
+	}
+	var rc ResourceConfig
+	if q, exists := rl[v1.ResourceMemory]; exists {
+		// Memory is defined in bytes.
+		val := q.Value()
+		rc.Memory = &val
+	}
+	if q, exists := rl[v1.ResourceCPU]; exists {
+		// CPU is defined in milli-cores.
+		val := MilliCPUToShares(q.MilliValue())
+		rc.CpuShares = &val
+	}
+	return &rc
+}
+
+// getNodeAllocatableInternal returns the Node Allocatable based on the purpose.
+// Acceptible values for purpose are `forScheduling` and `forEnforcement`.
+// If `forScheduling` is the purpose, then hard eviction thresholds are subtracted from capacity while computing Node Allocatable.
+// If `forEnforcement` if the purpose, then hard eviction thresholds are ignored.
+// Returns a ResourceList.
+func (cm *containerManagerImpl) getNodeAllocatableInternal(purpose int) v1.ResourceList {
```

Comment:
feels like we should have an enum of some kind... the int feels a bit open ended here...